echopype/calibrate/ek80_complex.py

```diff
@@ -273,47 +305,41 @@ def compress_pulse(backscatter: xr.DataArray, chirp: Dict) -> xr.DataArray:
     xr.DataArray
         A data array containing pulse compression output.
     """
-    pc_all = []
-
-    for chan in backscatter["channel"]:
-        # Select channel `chan` and drop the specific beam dimension if all of the values are nan.
-        backscatter_chan = backscatter.sel(channel=chan).dropna(dim="beam", how="all")
-
-        # Create NaN mask
-        # If `backscatter_chan` is lazy loaded, then `nan_mask` too will be lazy loaded.
-        nan_mask = np.isnan(backscatter_chan)
-
-        # Zero out backscatter NaN values
-        # If `nan_mask` is lazy loaded, then resulting `backscatter_chan` will be lazy loaded.
-        backscatter_chan = xr.where(nan_mask, 0.0 + 0j, backscatter_chan)
-
-        # Extract transmit values
-        tx = chirp[str(chan.values)]
-
-        # Compute complex conjugate of transmit values and reverse order of elements
-        replica = np.flipud(np.conj(tx))
-
-        # Apply convolve on backscatter (along range sample dimension) and replica
-        pc = xr.apply_ufunc(
-            lambda m: (signal.convolve(m, replica, mode="full")[replica.size - 1 :]),
-            backscatter_chan,
-            input_core_dims=[["range_sample"]],
-            output_core_dims=[["range_sample"]],
-            dask="parallelized",
-            vectorize=True,
-            output_dtypes=[np.complex64],
-        ).compute()
-
-        # Restore NaN values in the resulting array.
-        # Computing of `nan_mask` here is necessary in the case when `nan_mask` is lazy loaded
-        # or else the resulting `pc` will also be lazy loaded.
-        pc = xr.where(nan_mask.compute(), np.nan, pc)
+    # Calculate the transmit signal values from the chirp dictionary
+    replica_dict = {
+        # Compute conjugate and flip for each channel's transmit signal
+        str(channel.values): np.flipud(np.conj(chirp[str(channel.values)]))
+        for channel in backscatter["channel"]
+    }
+
+    # Zero out backscatter NaN values
+    nan_mask = np.isnan(backscatter)
+    backscatter_with_zeroed_nans = xr.where(nan_mask, 0.0 + 0.0j, backscatter)
+
+    # Create a partial function of the convolve function to pass in chirp and channels
+    _convolve_per_channel_partial = partial(
+        _convolve_per_channel,
+        replica_dict=replica_dict,
+        channels=backscatter_with_zeroed_nans["channel"],
+    )
 
-        pc_all.append(pc)
+    # Apply convolve on backscatter and replica (along range sample and channel dimension):
+    # To enable parallelized computation with `dask='parallelized'`, we rechunk to ensure that
+    #  the data is chunked with only one chunk along the core dimensions.
+    pc = xr.apply_ufunc(
+        _convolve_per_channel_partial,
+        backscatter_with_zeroed_nans.chunk({"range_sample": -1, "channel": -1}),
```

Comment:
Here I wonder if it makes difference if you make one chunk per channel, like how you did it in the noise filter case.

Reply:
Since it's a core dimension in `apply_ufunc`, I have to chunk it into a single chunk. It is enforced in this case.

Reply:
Ah I see, ok. Thanks! We can see how things look with large datasets when we get to that.